GuessTheNumber: Add initial version (WIP)

gh215 opened the pull request.

gh215 committed:
GuessTheNumber: Add initial version

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,71 @@
+
+        computerNumber = rand() % (max - min + 1) + min;
+
+        cout << "Компьютер загадал такое число: " << computerNumber << ". Оно больше, меньше или равно загаданному числу?" << endl;
```
У пользователя нет информации о том, какими числами следует давать ответ.

amavrin replied:
Давай переносить динные строчки
```suggestion
        cout << "Компьютер загадал такое число: " 
          << computerNumber 
          << ". Оно больше, меньше или равно загаданному числу?" 
          << endl;
```

gh215 committed:
Update chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp

Co-authored-by: Alexey Mavrin <amavrin@gmail.com>

gh215 committed:
GuessTheNumber: Code update to avoid the "integer divided by zero" error

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,83 @@
+       
+    do
+    {
+       if ((max > 0 && min > 0) && (max < 51 && min < 51))
```
Использование "магических чисел" 0, 1, 50  и 51 - неудачная практика.
У тебя есть переменные min и max, их, например можно использовать.

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,83 @@
+       {
+           cout << "Ошибка! Загаданное число не может быть 0 или больше 50!" << endl;
+           return 0;
+       }
```
Было бы понятнее, если бы короткая ветка `if - else` располагалась сначала.

Сейчас у тебя так:

```
if (условие)
{
  много
  строчек
  кода
}
else
{
  одна строка
  return
}
```

Если сделать наоборот - будет проще читать:
```
if (!условие)
{
  одна строка
  return
}
много
строчек
кода
```

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,83 @@
+    если число 0 - ошибка
+    если число больше 50 - ошибка
+    если число = 1 - минимальное значение
+    если число = 50 - максимальное значение
```
Я попробовал запустить программу и получил такое:
```
Вы будете писать 1, 2 или 3 после каждого числа, выведенного компьютером:
1: Число слишком маленькое
2: Число слишком большое
3: Число было угадано
Компьютер загадал такое число: 8. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2
Число слишком большое
Компьютер загадал такое число: 16. Оно больше, меньше или равно загаданному числу?
Ваш ответ:
```
То есть компьютер предложил 8, я ответил, что это много, и компьютер, вместо того, чтобы предложить число меньшее 8, наоборот, предложил число большее 8. Это логическая ошибка.

Или вот - что-то странное:
```
Вы будете писать 1, 2 или 3 после каждого числа, выведенного компьютером:
1: Число слишком маленькое
2: Число слишком большое
3: Число было угадано
Компьютер загадал такое число: 8. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2
Число слишком большое
Компьютер загадал такое число: 16. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 11. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 9. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 1144108939. Оно больше, меньше или равно загаданному числу?
Ваш ответ:
```
Компьютер предложил число `1144108939`

amavrin reviewed (changes requested):
Я предлагаю тебе написать тест-кейсы.

Вот как может выглядеть пример:

```
пользователь загадывает 1

повторяется
* если компьютер предлагает 1 - подтвердить правильность, конец программы.  компьютер
* предлагает число между 1 и 50
* каждое следующее предложенное число должно быть меньше предыдущего.
* пользователь отвечает - "большое"
```

gh215 committed:
GuessTheNumber: Updated test-cased and code update

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,84 @@
+    пользователь отвечает - "большое" - число уменьшается
+    пользователь отвечает - "маленькое" - число увеличивается
+    если компьютер предлагает 50 - подтвердить правильность, конец программы.
+    если пользователь отвечает - "большое" - ошибка
```
Если бы я был тестировщиком - я бы ничего не понял 0-:E

amavrin reviewed (changes requested):
Вижу странное поведение компьютера:
```
Компьютер загадал такое число: 11. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 42. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 42. Оно больше, меньше или равно загаданному числу?
```
тут компьютер не увеличил число, хотя должен был бы.

А тут вообще что-то странное:
```
1: Число слишком маленькое 
2: Число слишком большое 
3: Число было угадано
Компьютер загадал такое число: 50. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 49. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 49. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 48. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 49. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 48. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 47. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 46. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 45. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 45. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 48. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 50. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1
Число слишком маленькое
Компьютер загадал такое число: 50. Оно больше, меньше или равно загаданному числу?
```

В общем - нужно разбираться.
Попробуй, как всегда, начать с бумажки. Нарисуй числовую ось, отметь макс и мин, и поиграй сам с собой. Тебе необходимо до конца уяснить, какой алгоритм действий должен быть у компьютера.

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,84 @@
+           cout << "Ошибка! Загаданное число не может быть 0 или больше 50!" << endl;
+           return 0;
+       }
+       {
```
А зачем тут открыт новый блок кода (скобкой `{`)?

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,84 @@
+    int min = 1;
+    int max = 50;
+    const int limit_max_Number = 51;
+    const int limit_min_Number = 0;
```
Почему 51 и 0? Как они связаны с min = 1 и max = 50?

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,84 @@
+    const int limit_min_Number = 0;
+    srand(static_cast<unsigned int>(time(0)));
+
+    cout << "Добро пожаловать в нашу игру! Вы загадываете число в уме от 1 до 50, а компьютер его отгадывает"
```
Тут лучше использовать min и max, а не фиксированные числа

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,84 @@
+           {
+               cout << "Компьютер угадал ваше число и оно = " << computerNumber <<
+                   " и на это ушло " << tries << " попыток" << endl;
+               return 1;
```
Возврат числа, отличного от 0 из main() обычно означает ошибку, так что тут не нужно возвращать 1. Лучше вернуть 0

gh215 committed:
GuessTheNumber: Start fixing test-cases; fix formatting

gh215 committed:
GuessTheNumber: added implementation of test-cases of abnormal user: continuation of the game after wrong number set and
declaring the user as a dishonest player if he changes the set number

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+    число меньше - К генерирует число больше
+    3. П задумал число на середине:
+    Сужает диапазон. Число меньше - ставит меньшую границу диапазона
+    Число больше - ставит большую границу диапазона
```
Тест-кейсы по-прежнему недостаточно проработаны.

Во-первых, в них ничего не сказано про повторения. Можно подумать, что если ты один раз ответил "2", а К. уменьшил свой guess, то всё ОК. На самом деле - К. должен пытаться, пока не угадает.

Дальше: представь себе, что ты тестировщик, и тестируешь кейс 1.

Ты задумал число на нижней границе, и отвечаешь всегда `2`.
К. предлагает тебе числа 40, 30, 20, 10, 0.
По тест-кейсу формально проходит.
А по сути - нет, поскольку К не угадал число

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+    Тест-кейсы для ненормального пользователя
+    1. За диапазоном - выдаёт ошибку
+    2. Недопустимы ответ (например 4), компьютер должен продолжать игру
+    3. Пользователь меняет показания - программа называет его обманщиком
```
Это пока просто похоже на намётки, которые ты сделал себе, чтобы не забыть, но не являются проработанным тест-кейсами.

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+#include <string>
+
+
+/* Тест-кейсы для нормального пользователя
```
Я задумал 40, и получил следующее:

```
Компьютер загадал такое число: 1. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое

Компьютер загадал такое число: 39. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое

Компьютер загадал такое число: 42. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2

Ответ пользователя: число слишком большое

Компьютер загадал такое число: 41. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2

Ответ пользователя: число слишком большое

Ошибка! Загаданное число не может быть 0 или больше 50!
```

К не угалал, но выдал ошибку.
В чём проблема?

amavrin reviewed (changes requested):
Нужно доработать тест-кейсы и найти проблему, которая привела к ошибке

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+            cout << "\nОшибка! Неверно введённое число!" << endl;
+            continue;
+        }      
+        if (min > max || max < min)
```
min > max и max < min - это одно и то же 😉

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+    int tries = 0;
+    int min = 1, max = 50;
+
+    bool changing_numbers = false;
```
Переменная `changing_numbers` не используется

amavrin commented on chapter-2/GuessTheNumber/GuessTheNumber/GuessTheNumber.cpp:
```diff
@@ -0,0 +1,105 @@
+            return 0;
+        }
+       
+    }   while (user_conceive != CORRECT && tries != 7); 
```
А зачем ты ввёл максимальное число попыток компьютера?
Ограничение имеет значение при тестировании - при корректной работе количество попыток в среднем будет равно логарифму по основанию 2 от длины диапазона от min до max, но для нормально работающей программы эта проверка не нужна.

Кроме того, это привело к использованию "магического числа" 7.

gh215 committed:
GuessTheNumber: Finalized and fixed 3 test case of abnormal user.
Implemented the computer checking 1 and 50 as bounds of numbers and informing
the user that this is the maximum and minimum number

amavrin reviewed (changes requested):
Поведение программы некорректно. Вот моя сессия:

```
Вы будете писать 1, 2 или 3 после каждого числа, выведенного компьютером: 
1: Число слишком маленькое 
2: Число слишком большое 
3: Число было угадано

Компьютер загадал такое число: 26. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое

Компьютер загадал такое число: 28. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2

Ответ пользователя: число слишком большое

Компьютер загадал такое число: 27. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое
Floating point exception (core dumped)
```

Что произошло:
- допустим, П загадал 27.
- К предложил 26 - П ответил "1"
- К предложил 28 - П ответил "2"
- если 26 мало, а 28 - много - единственный вариант 27. Но К продолжает спрашивать. Это первая некорректность
- П ответил, что число слишком маленькое; К упал с ошибкой.

Чтобы не казалось, что это случайность - вот ещё подобная ситуация:

```
Вы будете писать 1, 2 или 3 после каждого числа, выведенного компьютером: 
1: Число слишком маленькое 
2: Число слишком большое 
3: Число было угадано

Компьютер загадал такое число: 31. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое

Компьютер загадал такое число: 38. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2

Ответ пользователя: число слишком большое

Компьютер загадал такое число: 33. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 2

Ответ пользователя: число слишком большое

Компьютер загадал такое число: 32. Оно больше, меньше или равно загаданному числу?
Ваш ответ: 1

Ответ пользователя: число слишком маленькое
Floating point exception (core dumped)
```

На мой взгляд, нужно изменить подход к написанию данной программы.
Самое первое, что нужно сделать - это хорошо осознать алгоритм, по которому должна работать программа. Постарайся как следует осознать его для себя я затем описать его в виде блок-схемы. Будем стараться разработать корректную программу. Описание корректного алгоритма - первый шаг.
